libraries/adaptive-expressions/src/triggerTrees/trigger.ts

```diff
@@ -0,0 +1,502 @@
+import { PredicateComparer } from './optimizer';
+import { Quantifier, QuantifierType } from './quantifier';
+import { RelationshipType } from './relationshipType';
+import { TriggerTree } from './triggerTree';
+
+/**
+ * Rewrite the expression by pushing not down to the leaves.
+ */
+const pushDownNot = (expression: Expression, inNot = false): Expression => {
+    let newExpr = expression;
+    const negation = expression.evaluator.negation;
+    switch (expression.type) {
+        case ExpressionType.And:
+        case ExpressionType.Or:
+            const children = expression.children.map(child => pushDownNot(child, inNot));
+            if (children.length === 1) {
+                newExpr = children[0];
+            } else {
+                newExpr = Expression.makeExpression(
+                    expression.type === ExpressionType.And
+                        ? (inNot ? ExpressionType.Or : ExpressionType.And)
+                        : (inNot ? ExpressionType.And : ExpressionType.Or),
+                    undefined,
+                    ...children);
+            }
+            break;
+        case ExpressionType.Not:
+            newExpr = pushDownNot(expression.children[0], !inNot);
+            break;
+        default:
+            if (inNot) {
+                if (negation) {
+                    if (expression.type === negation.type) {
+                        // Pass through like optional/ignore
+                        newExpr = Expression.makeExpression(
+                            undefined,
+                            negation,
+                            ...expression.children.map(child => pushDownNot(child, true))
+                        );
+                    } else {
+                        // Replace with negation and stop
+                        newExpr = Expression.makeExpression(
+                            undefined,
+                            negation,
+                            ...expression.children
+                        );
+                    }
+                } else {
+                    // Keep not
+                    newExpr = Expression.makeExpression(ExpressionType.Not, undefined, expression);
+                }
+            }
+            break;
+    }
+
+    return newExpr;
+};
+
+/**
+ * A trigger is a combination of a trigger expression and the corresponding action.
+ */
+export class Trigger {
+    private readonly _quantifiers: Quantifier[];
+    private readonly _tree: TriggerTree;
+    private _clauses: Clause[];
+
+    /**
+     * Intializes a new instance of the `Trigger` class.
+     * @param tree Trigger tree that contains this trigger.
+     * @param expression Expression for when the trigger action is possible.
+     * @param action Action to take when a trigger matches.
+     * @param quantifiers Quantifiers to dynamically expand the expression.
+     */
+    public constructor(tree: TriggerTree, expression: Expression, action: any, ...quantifiers: Quantifier[]) {
```

Comment:
Actually the `action` can be anything, it can be a string or an object or something else, it is just a container for real "actions". e.g. In trigger tree tests, we use strings as actions. In MostSpecificSelector, we use `OnCondition` objects as actions.
When I was porting this class from C#, the type of action was `object` because anything can be object in c#. So I have to leave it `any` here.